Project/Source/Components/UI/ComponentText.cpp

```diff
@@ -187,32 +223,72 @@ void ComponentText::RecalculcateVertices() {
 
 	ComponentTransform2D* transform = GetOwner().GetComponent<ComponentTransform2D>();
 	float3 position = transform->GetPosition();
+	float screenFactor = GetOwner().GetComponent<ComponentCanvasRenderer>()->GetCanvasScreenFactor();
 
-	float x = position.x;
-	float y = position.y;
+	float x = position.x * screenFactor;
+	float y = position.y * screenFactor;
+	
+	float dy = 0;		// additional y shifting
+	int j = 0;			// index of row
 
 	float2 transformScale = transform->GetScale().xy();
-
-	float scale = (fontSize / 48) * (transformScale.x > transformScale.y ? transformScale.x : transformScale.y) * GetOwner().GetComponent<ComponentCanvasRenderer>()->GetCanvasScreenFactor();
+	// FontSize / size of imported font. 48 is due to FontImporter default PixelSize
+	float scale = (fontSize / 48) * (transformScale.x > transformScale.y ? transformScale.x : transformScale.y) * screenFactor;
 
 	for (int i = 0; i < text.size(); ++i) {
 		Character character = App->userInterface->GetCharacter(fontID, text.at(i));
+		char c = text.at(i);
 
 		float xpos = x + character.bearing.x * scale;
 		float ypos = y - (character.size.y - character.bearing.y) * scale;
 
 		float w = character.size.x * scale;
 		float h = character.size.y * scale;
 
+		switch (textAlignment) {
+			case TextAlignment::LEFT: {
+				// Default branch, could be deleted
+				break;
+			}
+			case TextAlignment::CENTER: {
+				xpos += (transform->GetSize().x * screenFactor / 2.0f - SubstringWidth(&text.c_str()[j], scale) / 2.0f);
+				break;
+			}
+			case TextAlignment::RIGHT: {
+				xpos += transform->GetSize().x * screenFactor - SubstringWidth(&text.c_str()[j], scale);
+				break;
+			}
+		}
+
+		if (text.at(i) == '\n') {
+			dy += lineHeight;					// shifts to next line
+			x = position.x * screenFactor;		// reset to initial position
+			j = i + 1;							// updated j variable in order to get the substringwidth of the following line in the next iteration
+		}
+
 		verticesText[i] = {
-			xpos, ypos + h, 0.0f, 0.0f, 
-			xpos, ypos,	0.0f, 1.0f,	
-			xpos + w, ypos, 1.0f, 1.0f,
-			xpos, ypos + h, 0.0f, 0.0f,
-			xpos + w, ypos, 1.0f, 1.0f,
-			xpos + w, ypos + h, 1.0f, 0.0f};
+			xpos, ypos + h - dy, 0.0f, 0.0f,
+			xpos, ypos - dy, 0.0f, 1.0f,
+			xpos + w, ypos - dy, 1.0f, 1.0f,
+			xpos, ypos + h - dy, 0.0f, 0.0f,
+			xpos + w, ypos - dy, 1.0f, 1.0f,
+			xpos + w, ypos + h - dy, 1.0f, 0.0f
+		};
 
 		// now advance cursors for next glyph (note that advance is number of 1/64 pixels)
-		x += (character.advance >> 6) * scale; // bitshift by 6 to get value in pixels (2^6 = 64)
+		if (text.at(i) != '\n') {
+			x += (character.advance >> 6) * scale; // bitshift by 6 to get value in pixels (2^6 = 64). Divides / 64
+		}
+	}
+}
+
+float ComponentText::SubstringWidth(const char* substring, float scale) {
+	float subWidth = 0.f;
+
+	for (int i = 0; i < substring[i] != '\0' && substring[i] != '\n'; ++i) {
```

Comment:
I don't understand what's happening here. Is this for correct?

Reply:
They are getting the line width (in "character units"). So if the string has a \0 or a \n, means a line break, then stop there and return that width.
Anyhow, I dont get the "i <" either

Reply:
Yeah, I'm removing the "i <". Comes from a deprecated chunk of code for an alternative version of the function. Missed it, but works correctly anyway.